### ISSUE TITLE
"QueryExecutionInfo#isSuccess" is reported as failure with "ReactiveCrudRepository#save"

### ISSUE THREAD

**Issue by ttddyy** [r2dbc/r2dbc-proxy#55]
From gitter discussion:
https://gitter.im/R2DBC/r2dbc?at=5e1c726865540a529a0c9366

I have reproduced the issue using H2 with `ReactiveCrudRepository#save`

```java
@GetMapping("/insert")
//    @Transactional
public Mono<City> insert() {
  City city = new City("name", "country");
  return this.repository.save(city)
    .doFinally(signal -> {
      System.out.println("CONTROLLER signal=" + signal);
    });
  };
```

The method are called in following sequence:
```
[ 33] [before-method] io.r2dbc.spi.ConnectionFactory#create
[ 34] [after-method] io.r2dbc.spi.ConnectionFactory#create
[ 35] [before-method] io.r2dbc.spi.Connection#createStatement
[ 36] [after-method] io.r2dbc.spi.Connection#createStatement
[ 37] [before-method] io.r2dbc.spi.Statement#bind
[ 38] [after-method] io.r2dbc.spi.Statement#bind
[ 39] [before-method] io.r2dbc.spi.Statement#bind
[ 40] [after-method] io.r2dbc.spi.Statement#bind
[ 41] [before-method] io.r2dbc.spi.Statement#returnGeneratedValues
[ 42] [after-method] io.r2dbc.spi.Statement#returnGeneratedValues
[ 43] [before-query] Query:["INSERT INTO city (name, country) VALUES ($1, $2)"]
[ 44] [before-method] io.r2dbc.spi.Statement#execute
[ 45] [before-method] io.r2dbc.spi.Result#map
[ 46] [before-method] io.r2dbc.spi.Connection#close
[ 47] [after-method] io.r2dbc.spi.Connection#close
CONTROLLER signal=onComplete
[ 48] [after-method] io.r2dbc.spi.Result#map
[ 49] [after-method] io.r2dbc.spi.Statement#execute
[ 50] [after-query] Query:["INSERT INTO city (name, country) VALUES ($1, $2)"]
```

The (34), 48, 49 receives cancel signal.

According to @mp911de

> save(…) calls either insert or update methods and takes the first() element of the result
> See https://github.com/spring-projects/spring-data-r2dbc/blob/master/src/main/java/org/springframework/data/r2dbc/repository/support/SimpleR2dbcRepository.java#L74-L97

cc/ @aravindtga @squiry

**Comment by ttddyy** [comment]
The `SimpleR2dbcRepository#save()`, in turn, calls [`DefaultFetchSpec#first()`](https://github.com/spring-projects/spring-data-r2dbc/blob/7ec3d28b17ef4c4a6c77b355c777cccf9519cb10/src/main/java/org/springframework/data/r2dbc/core/DefaultFetchSpec.java#L71-L74) which performs `all().next()`.

The `Flux#next()` does cancel the original source.

![Flux#next marble](https://projectreactor.io/docs/core/release/api/reactor/core/publisher/doc-files/marbles/next.svg)
(From [Flux#next](https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Flux.html#next).)

I think that's why `Statement#execute` received cancel signal.


Currently, `r2dbc-proxy` sets query is executed successful when `onComplete` is called on the publisher returned by `Statement#execute`. Therefore, this cancel signal will trigger not to call `onComplete` and ended up considering the insert to be failed(not setting `QueryExecutionInfo#success` to `true`).

@mp911de Is this canceling ideal behavior on spring-data-r2dbc side? How about using `one()` instead of `first()` on `SimpleR2dbcRepository#save()`?
I just locally tweaked the code and using `one()` properly completing the `Statement#execute` publisher.

**Comment by mp911de** [comment]
Let's take a step back to see why cancellation is part of the process.

By default, Publishers emit items until they finish doing so. For regular queries, where results are consumed as `Flux<T>`, this is perfectly fine. In a few other cases, we want to limit the results using either proper SQL (`SELECT TOP n …`, `INSERT`, `UPDATE`), limit the result stream (`take(n)`, `first()`) or both.

While SQL-based limitation is the best way to reduce the number of rows for `Flux`-based consumption, we have a few cases, where we want to limit results and switch to a `Mono`. Both, `one()` and `first()` (as of a Spring Data perspective) are used to do so. `first()` uses internally the `next()` operator to obtain a single item and then it cancels the upstream consumption.

So `next()` does not await the completion signal even if there is a single element being emitted by the upstream `Publisher`.
`one()` uses internally `buffer(2)`. If the first emitted buffer contains a single item, then the single item gets propagated, if the buffer contains two elements, then an exception is thrown. In consequence, the upstream `Publisher` gets canceled too.

Where I'm basically heading is that cancellation is part of the entire way of thinking. We cannot know whether `cancel()` was called because some downstream component experienced a failure or whether `cancel()` is to signal: Thanks, I've received a sufficient amount of data.

There are all sorts of issues associated with cancelation, such as we cannot await completion if the upstream was canceled. See also spring-projects/spring-framework#23304 for a related discussion.

I expect other client libraries to also utilize cancelation. For this particular case, I suggest that a query execution is considered successful if the subscription has seen either a completion signal or an element. For a query, where we haven't seen a data signal yet (because it was canceled), we cannot know whether it was successful or not.

**Comment by ttddyy** [comment]
ok.
@mp911de  Thanks for the explanation.

I'll update condition for query execution success to be either completion or at least one item is emitted. This should handle the cancelation triggered by `next()` usecase.

**Comment by ttddyy** [comment]
@gregturn @mp911de 
I have created a fix for this issue on [55-query-success branch](/r2dbc/r2dbc-proxy/tree/55-query-success).

What is the logistics for merging the fix to `0.8.1` release?
I see there is `0.8.x` branch. So, I assume push the commit to `0.8.x` branch and merge it back to `master` for `0.9.0` future release?

**Comment by mp911de** [comment]
Yeah, that's correct. We will cut a release from our 0.8.x branches. I learned that while merging (merge commit) a change to a single branch is simple, porting that change to another is something that messes entirely with my backport branch. So cherrypicking is probably the way to go.
